drivers/led/lp50xx.c

```diff
@@ -0,0 +1,408 @@
+
+#define LP50XX_MIN_BRIGHTNESS		0U
+#define LP50XX_MAX_BRIGHTNESS		100U
+
+#define LP5009_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5012_MAX_LEDS)
+#define LP5012_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5012_MAX_LEDS)
+#define LP5018_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5024_MAX_LEDS)
+#define LP5024_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5024_MAX_LEDS)
+#define LP5030_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5036_MAX_LEDS)
+#define LP5036_NUM_CHANNELS	((LP50XX_COLORS_PER_LED + 1) * LP5036_MAX_LEDS)
+
+/* Common registers */
+#define LP50XX_DEVICE_CONFIG0		0x00
+#define LP50XX_DEVICE_CONFIG1		0x01
+#define LP50XX_LED_CONFIG0		0x02
+
+/* LP5009/12 registers */
+#define LP5012_LED0_BRIGHTNESS		0x07
+#define LP5012_OUT0_COLOR		0x0b
+
+/* LP5018/24 registers */
+#define LP5024_LED0_BRIGHTNESS		0x07
+#define LP5024_OUT0_COLOR		0x0f
+
+/* LP5030/36 registers */
+#define LP5036_LED_CONFIG1		0x03
+#define LP5036_LED0_BRIGHTNESS		0x08
+#define LP5036_OUT0_COLOR		0x14
+
+/* Register values */
+#define CONFIG0_CHIP_EN			BIT(6)
+
+#define CONFIG1_LED_GLOBAL_OFF		BIT(0)
+#define CONFIG1_MAX_CURRENT_OPT		BIT(1)
+#define CONFIG1_PWM_DITHERING_EN	BIT(2)
+#define CONFIG1_AUTO_INCR_EN		BIT(3)
+#define CONFIG1_POWER_SAVE_EN		BIT(4)
+#define CONFIG1_LOG_SCALE_EN		BIT(5)
+
+/* Chipset models supported by this driver. */
+enum lp50xx_chips {
+	LP5009,
+	LP5012,
+	LP5018,
+	LP5024,
+	LP5030,
+	LP5036,
+};
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	enum lp50xx_chips chip;
+	uint8_t num_leds;
+	bool log_scale_en;
+	bool max_curr_opt;
+	const struct led_info *leds_info;
+};
+
+struct lp50xx_data {
+	uint8_t *chan_buf;
+};
+
+static int lp50xx_get_max_leds(enum lp50xx_chips id)
+{
+	switch (id) {
+	case LP5009:
+	case LP5012:
+		return LP5012_MAX_LEDS;
+	case LP5018:
+	case LP5024:
+		return LP5024_MAX_LEDS;
+	case LP5030:
+	case LP5036:
+		return LP5036_MAX_LEDS;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lp50xx_get_num_channels(enum lp50xx_chips id)
+{
+	switch (id) {
+	case LP5009:
+		return LP5009_NUM_CHANNELS;
+	case LP5012:
+		return LP5012_NUM_CHANNELS;
+	case LP5018:
+		return LP5018_NUM_CHANNELS;
+	case LP5024:
+		return LP5024_NUM_CHANNELS;
+	case LP5030:
+		return LP5030_NUM_CHANNELS;
+	case LP5036:
+		return LP5036_NUM_CHANNELS;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lp50xx_get_brightness_base_addr(enum lp50xx_chips id)
+{
+	switch (id) {
+	case LP5009:
+	case LP5012:
+		return LP5012_LED0_BRIGHTNESS;
+	case LP5018:
+	case LP5024:
+		return LP5024_LED0_BRIGHTNESS;
+	case LP5030:
+	case LP5036:
+		return LP5036_LED0_BRIGHTNESS;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lp50xx_get_color_base_addr(enum lp50xx_chips id)
+{
+	switch (id) {
+	case LP5009:
+	case LP5012:
+		return LP5012_OUT0_COLOR;
+	case LP5018:
+	case LP5024:
+		return LP5024_OUT0_COLOR;
+	case LP5030:
+	case LP5036:
+		return LP5036_OUT0_COLOR;
+	default:
+		return -EINVAL;
+	}
+}
```

Comment:
The only information known at build time is the chip id which gets added to the `lp50xx_config` structure for each device present in the devicetree. Then, the associated register will only be known at runtime since we need to get the value of `dev->config->chip`.

I can replace all the functions with macros with ternary operators like that : 
`#define LP50XX_MAX_LEDS(chip)				\`
`	((chip) <= LP5012) ? LP5012_MAX_LEDS :		\`
`	((chip) <= LP5024) ? LP5024_MAX_LEDS :		\`
`	((chip) <= LP5036) ? LP5036_MAX_LEDS :		\`
`	-EINVAL									 `

but won't inlining the function have the same effect ? Just tell me what you think and I can make the change like above for all functions if you find it better.

Reply:
Yes you are right. Actually your first approach (i.e. storing this information in `struct lp50xx_config`) was good. I asked you to not waste memory and I said:

> These values ​​are shared/duplicated between all LEDs

I don't know what I was thinking and I was wrong. The cost is only a few bytes per chip in `struct lp50xx_config` (and not per LED) and it was way better than the conditions we are adding in the code now. Please roll back to your previous version for this part. I apologize for the inconvenience.

Nevertheless check that all the fields you add are needed. For example I am not sure for `color_base_reg`. If we know the number of channels and the brightness base register then it can be computed easily. Please double check for the other fields too.